Source/missiles.cpp

```diff
@@ -915,6 +874,105 @@ Direction16 GetDirection16(Point p1, Point p2)
 	return ret;
 }
 
+void MissileHitMonsterConsequences(int mid, int pnum, int dam, missile_id mName)
+{
+	auto &monster = Monsters[mid];
+	const auto &player = Players[pnum];
+
+	bool hasKnockback = false;
+	if (pnum >= 0) {
+		const auto &player = Players[pnum];
+		if (HasAnyOf(player._pIFlags, ItemSpecialEffect::Knockback))
+			hasKnockback = true;
+	}
+
+	if (monster._mhitpoints >> 6 <= 0) {
+		M_StartKill(mid, pnum);
+	} else if (monster.IsResistant(mName)) {
+		PlayEffect(monster, 1);
+	} else {
+		if (monster._mmode != MonsterMode::Petrified
+		    && MissilesData[mName].mType == 0
+		    && hasKnockback)
+			M_GetKnockback(mid);
+		if (monster.MType->mtype != MT_GOLEM)
+			M_StartHit(mid, pnum, dam);
+	}
+}
+
+int PlayerMissileCalculateCTHAgainstMonster(int pnum, missile_id mName, int dist, Monster &monster)

```

Comment:
`Player` Considering that the function also accepts not being given a valid player id this function doesn't seam specific to players.

`AgainstMonster` is given by the last argument. This seams to be a good case for overloading where there could be a similar function that takes Player as the last argument.

Maybe a better name is `CalculateMissileChanceToHit()`, or even `Missile::CalculateChanceToHit()` so that it matches `Player::CalculateArmorPierce()` which it calls in turn.

Reply:
1. MonsterMHit is called twice from CheckMissileCol
-  `_micaster == TARGET_MONSTERS` and `!missile.IsTrap()`
- `_micaster == TARGET_BOTH`

first - missile from Player only.
second - Walls, Runes, FireRing and Teleport ;)

I am not 100% sure, but 95% that this is Player's only, and even if there are rare cases - MonsterMHit acts like it would be called by Player Missile. (regardless if pnum = -1 exception. Used or not).

There is also MonsterTrapHit called, when mosnters missile hitting golem - and I will put it to TrapMissile class, because it is more about that it acts like a trap.

Let's get back to naming discussion after commit with 2 functions merge to 1 in this PR.

2. Yes I was thinking this may end like that finally. After `Plr2PlrMHit` and `PlayerMHit` refactor I plan, when this one would be succefully merged, those functions could be overloaded. But I will go with your suggestion now.

Reply:
It turned out that the only exception was Ring of Fire as trap.
I moved this Hellfire exclusive logic out from `PlayerMissile::CalculateCTH` to `TrapMissile::CalculateCTH`.
Now all calls equivalents to MonsterMHit shall be treated as PlayerMissile.